### PR TITLE
[Refactor] Improve CI Run Times by Reducing Test Flakiness and Speeding up Test Suite

### PR DESCRIPTION
**Pull request by N2D4**
### Context

Our CI has evinced a few flakey tests recently. Additionally, the `freestyle-prod` action takes a long time to run. These end up delaying the merging/iterative development processes. There was also a bug with the existing implementation of `getExternalDbSyncFusebox` where it was performing an upsert, which according to [prisma](https://www.prisma.io/docs/orm/reference/prisma-client-reference#unique-key-constraint-errors-on-upserts), could cause conflicts. `getExternalDbSyncFusebox` was trying to upsert the settings singleton for the fusebox lazily if it wasn't updated yet, and if multiple upserts are occurring at once it can cause a rare unique key constraint violation on the singleton settings key. The fix is to not have the get function attempt an update.

---

> [!NOTE]
> **Medium Risk**
> Touches internal external-db-sync control flow and test/CI orchestration; behavior changes could affect how quickly sync-driven tests converge, though changes are scoped to internal endpoints and test harnesses.
> 
> **Overview**
> Reduces E2E/CI flakiness by removing the test-only “force external DB sync” mechanism and the `stopWhenIdle` query behavior from the internal external-db-sync `sequencer`/`poller`, so sync runs consistently for the full `maxDurationMs` (with added lightweight logging when work is found).
> 
> Stabilizes email-related tests by centralizing outbox helpers in `backend-helpers`, switching tests from fixed sleeps to polling for outbox status, and improving failure diagnostics by including outbox state in assertion errors. CI workflows are updated accordingly (drop `STACK_FORCE_EXTERNAL_DB_SYNC`, and adjust prod PR runs to execute the `mail` test subset).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 7bed7c57c87906e5666f02320bd84d01ecf0dd2c. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### Additional Notes
Note that we also update the Action for `freestyle-prod` to only run the mail tests on non-dev and non-main branches. This is because freestyle is used for email-rendering, and so, those tests are the ones we're chiefly concerned about. Running `freestyle-prod` tests for non-mail tests broadly doesn't reveal any information that we wouldn't get from the `freestyle-mock` tests and the mail tests. However, to account for the edge cases where it would, we let it run to completion on dev and main.

## Summary by CodeRabbit

* **Chores**
  * Removed forced external DB sync flag from CI workflows; adjusted test invocation to append a "mail" argument in certain prod freestyle runs.
* **Refactor**
  * Removed idle-stop behavior from synchronization services for steadier, duration-driven operation.
* **Tests**
  * Consolidated outbox/email helpers and standardized status-based waits; replaced fixed sleeps with polling.
  * Removed manual force-sync triggers, introduced marker/condition verification and increased several test timeouts.
  * Swapped one OTP sign-in precondition for a fast signup in a user test.

### CODE REVIEW

File: .github/workflows/e2e-api-tests.yaml [hunk 1]
```diff
@@ -160,7 +159,7 @@ jobs:
         run: sleep 10
 
       - name: Run tests
-        run: pnpm test run ${{ matrix.freestyle-mode == 'prod' && '--min-workers=1 --max-workers=1' || '' }}
+        run: pnpm test run ${{ matrix.freestyle-mode == 'prod' && '--min-workers=1 --max-workers=1' || '' }} ${{ matrix.freestyle-mode == 'prod' && github.ref != 'refs/heads/main' && github.ref != 'refs/heads/dev' && 'mail' || '' }}
```
[the remark this answers is not in the document]
Comment (nams1570) [reply .github/workflows/e2e-api-tests.yaml:162]: Intentional to reduce CI time for anything not dev/main

File: apps/e2e/tests/backend/backend-helpers.ts [hunk 2]
```diff
@@ -405,7 +435,11 @@ export namespace Auth {
         }
         await wait(100 + i * 20);
         if (i >= 30) {
-          throw new StackAssertionError(`Sign-in code message not found after ${i} attempts`, { response, messages: messages.map(m => ({ ...m, body: m.body && omit(m.body, ["html"]) })) });
+          throw new StackAssertionError(`Sign-in code message not found after ${i} attempts`, {
+            response,
+            messages: messages.map(m => ({ ...m, body: m.body && omit(m.body, ["html"]) })),
+            outboxEmails: await getOutboxEmails(),
```
[the remark this answers is not in the document]
Comment (nams1570) [reply apps/e2e/tests/backend/backend-helpers.ts:458]: Yeah, in a test environment this will not be an issue. `getOutboxEmails` will fail when either the backend isn't responding, there's a timeout issue (which is uncommon on localhost), or some other catastrophic issue. When these things happen, the test would be doomed anyway - the outbox email fetch failure would probably be more interesting to look at than the "sign in attempt failed"!

[the remark this answers is not in the document]
Comment (nams1570) [reply apps/e2e/tests/backend/backend-helpers.ts:458]: Yeah, in a test environment this will not be an issue. `getOutboxEmails` will fail when either the backend isn't responding, there's a timeout issue (which is uncommon on localhost), or some other catastrophic issue. When these things happen, the test would be doomed anyway - the outbox email fetch failure would probably be more interesting to look at than the "sign in attempt failed"!

File: apps/e2e/tests/backend/backend-helpers.ts [hunk 3]
```diff
@@ -186,6 +186,36 @@ export async function bumpEmailAddress(options: { unindexed?: boolean } = {}) {
...
+
+// Helper to poll the outbox until an email with the expected subject and status appears
+export async function waitForOutboxEmailWithStatus(subject: string, status: string) {
+  const maxRetries = 24;
+  let emails: any[] = [];
+  for (let i = 0; i < maxRetries; i++) {
+    emails = await getOutboxEmails({ subject });
+    if (emails.length > 0 && emails[0].status === status) {
+      return emails;
+    }
```
[the remark this answers is not in the document]
Comment (nams1570) [reply apps/e2e/tests/backend/backend-helpers.ts:227]: Fair comment, check followup commit.

File: apps/e2e/tests/backend/backend-helpers.ts [hunk 4]
```diff
@@ -186,6 +186,36 @@ export async function bumpEmailAddress(options: { unindexed?: boolean } = {}) {
...
+  return listResponse.body.items;
+}
+
+// Helper to poll the outbox until an email with the expected subject and status appears
+export async function waitForOutboxEmailWithStatus(subject: string, status: string) {
+  const maxRetries = 24;
+  let emails: any[] = [];
+  for (let i = 0; i < maxRetries; i++) {
+    emails = await getOutboxEmails({ subject });
+    if (emails.length > 0 && emails[0].status === status) {
```
[the remark this answers is not in the document]
Comment (nams1570) [reply apps/e2e/tests/backend/backend-helpers.ts:225]: This is intentional to ensure stale emails don't flag false positives. Ordering is ensured. See comment in updated commit for more details.

File: apps/e2e/tests/backend/backend-helpers.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nams1570) [reply apps/e2e/tests/backend/backend-helpers.ts]: This is intentional to ensure stale emails don't flag false positives. Ordering is ensured. See comment in updated commit for more details.

File: apps/e2e/tests/backend/endpoints/api/v1/external-db-sync-race.test.ts [hunk 5]
```diff
@@ -376,24 +373,25 @@ describe.sequential('External DB Sync - Race Condition Tests', () => {
...
+                [`${dbName}@example.com`],
+              );
+              if (res.rows.length !== 1) return false;
+              row = res.rows[0];
+              return true;
+            },
+            {
+              description: 'waiting for sync to complete',
+              timeoutMs: 10000,
+            },
```
[the remark this answers is not in the document]
Comment (nams1570) [reply apps/e2e/tests/backend/endpoints/api/v1/external-db-sync-race.test.ts:400]: See followup commit, this has been resolved.

File: apps/e2e/tests/backend/endpoints/api/v1/external-db-sync-race.test.ts [hunk 6]
```diff
@@ -376,24 +373,25 @@ describe.sequential('External DB Sync - Race Condition Tests', () => {
...
           );
```
[the remark this answers is not in the document]
Comment (nams1570) [reply apps/e2e/tests/backend/endpoints/api/v1/external-db-sync-race.test.ts]: See followup commit to see how we refactored this test.

File: apps/e2e/tests/backend/endpoints/api/v1/external-db-sync-race.test.ts [hunk 7]
```diff
@@ -376,24 +373,25 @@ describe.sequential('External DB Sync - Race Condition Tests', () => {
...
-            `
-              SELECT "display_name"
-              FROM "users"
-              WHERE "primary_email" = $1
-            `,
-            [`${dbName}@example.com`],
+          let row: any;
+          await waitForCondition(
+            async () => {
+              const res = await externalClient.query<{
```
[the remark this answers is not in the document]
Comment (nams1570) [reply apps/e2e/tests/backend/endpoints/api/v1/external-db-sync-race.test.ts]: See followup commit for fix.

File: apps/backend/src/app/api/latest/internal/external-db-sync/poller/route.ts [hunk 8]
```diff
@@ -258,8 +249,8 @@ export const GET = createSmartRouteHandler({
           const pendingRequests = await claimPendingRequests();
           iterationSpan.setAttribute("stack.external-db-sync.pending-count", pendingRequests.length);
 
-          if (stopWhenIdle && pendingRequests.length === 0) {
-            return { stopReason: "idle", processed: 0 };
+          if (pendingRequests.length > 0) {
+            console.log(`[Poller] Processing ${pendingRequests.length} pending requests`);
           }
```
[the remark this answers is not in the document]
Comment (nams1570) [reply apps/backend/src/app/api/latest/internal/external-db-sync/poller/route.ts]: removed in followup commit, was used to debug.

File: apps/backend/src/app/api/latest/internal/external-db-sync/sequencer/route.ts [hunk 9]
```diff
@@ -145,6 +138,10 @@ async function backfillSequenceIds(batchSize: number): Promise<boolean> {
 
     span.setAttribute("stack.external-db-sync.did-update", didUpdate);
 
+    if (didUpdate) {
+      console.log(`[Sequencer] Backfilled USR=${projectUserTenants.length} CC=${contactChannelTenants.length} DEL=${deletedRowTenants.length}`);
+    }
```
[the remark this answers is not in the document]
Comment (nams1570) [reply apps/backend/src/app/api/latest/internal/external-db-sync/sequencer/route.ts]: removed in followup commit, used to debug.

File: apps/e2e/tests/backend/endpoints/api/v1/external-db-sync-race.test.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (nams1570) [reply apps/e2e/tests/backend/endpoints/api/v1/external-db-sync-race.test.ts]: This test was refactored in a followup commit to better test the prescribed behavior,

File: apps/e2e/tests/backend/backend-helpers.ts [hunk 10]
```diff
@@ -186,6 +186,36 @@ export async function bumpEmailAddress(options: { unindexed?: boolean } = {}) {
...
+      return emails;
+    }
+    await wait(500);
```
[the remark this answers is not in the document]
Comment (nams1570) [reply apps/e2e/tests/backend/backend-helpers.ts]: Added a comment to explain the context. This ensures we don't get a false positive
